### PR TITLE
Moving up some parameter checks in the class hierarchy

### LINKED ISSUE #3282: Gates do not check if complex inputs passed in
### Information

- **Qiskit Terra version**: master
- **Python version**:
- **Operating system**:

### What is the current behavior?
This works:
```python

from qiskit import *

qc = QuantumCircuit(2, 2)
qc.h(0)
qc.rx(1+1j, 0)
qc.cx(0, 1)
qc.measure([0, 1], [0, 1])

qc.draw()
```

but it should not.  It only gets caught later with:

```
TypeError: can't convert complex to float
```

### PR DESCRIPTION
**Pull request by 1ucian0**
Fixes #3282 

@nonhermitian notices [that parameters are too broadly checked]( https://github.com/Qiskit/qiskit-terra/issues/3282#issuecomment-554943493). I moved some of the checks up the class hierarchy. This allows to be more prices in the kind of check and normalization for each class. Also, it can be improved further with a better class hierarchy for `Instructions`. But that's story for another time.

All the tests are passing. If reviewers find a uncovered case, let's add the test!

### CODE REVIEW

File: qiskit/circuit/instruction.py [hunk 1]
```diff
@@ -138,71 +138,58 @@ def params(self):
...
-
-                self._params.append(single_param.sym())
-            # example: u3(0.1, 0.2, 0.3)
-            elif isinstance(single_param, (int, float)):
-                self._params.append(single_param)
-            # example: Initialize([complex(0,1), complex(0,0)])
-            elif isinstance(single_param, complex):
-                self._params.append(single_param)
-            # example: snapshot('label')
-            elif isinstance(single_param, str):
```
Comment (ewinston) [inline qiskit/circuit/instruction.py]: Is this case still handled?

Comment (1ucian0) [reply qiskit/circuit/instruction.py]: strings in snapshots?

There are several tests running `circuit.snapshot('something')` and they pass.

File: qiskit/circuit/instruction.py [hunk 2]
```diff
@@ -137,29 +137,19 @@ def params(self):
...
-                raise CircuitError("invalid param type {0} in instruction "
-                                   "{1}".format(type(single_param), self.name))
+            self._params.append(self.normalize_parameter(single_param))
+
+    def normalize_parameter(self, parameter):
+        if isinstance(parameter, (ParameterExpression, int, float, complex, str, list)):
+            return parameter
+        # example: numpy.array([[1, 0], [0, 1]])
+        elif isinstance(parameter, numpy.ndarray):
+            self._params.append(parameter)
```
Comment (mtreinish) [inline qiskit/circuit/instruction.py:147]: Won't this append `None`?

File: qiskit/circuit/gate.py [hunk 3]
```diff
@@ -219,3 +219,13 @@ def broadcast_arguments(self, qargs, cargs):
...
+
+    def normalize_parameter(self, parameter):
+        # examples:
+        #  u3(0.1, 0.2, 0.3)
+        #  u2(pi/2, sin(pi/4))
+        if isinstance(parameter, (int, float)):
+            return parameter
+
+        else:
+            return super().normalize_parameter(parameter)
```
Comment (mtreinish) [inline qiskit/circuit/gate.py:231]: I'm confused how is this different from the parent `Instruction` class? If parameter is a list str complex or parameter expression this just will return parameter too

File: qiskit/circuit/instruction.py (hunk 2 above)
Comment (1ucian0) [reply qiskit/circuit/instruction.py:147]: sorry.. I'm still merging this.

File: qiskit/circuit/gate.py (hunk 3 above)
Comment (1ucian0) [reply qiskit/circuit/gate.py:231]: sorry.. I'm still merging this.

Comment (1ucian0) [reply qiskit/circuit/gate.py:231]: You are right. Removed in c5b7fe819.

File: qiskit/circuit/instruction.py (hunk 2 above)
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: Sorry.. it was a bad merged. Fixed.

File: qiskit/circuit/instruction.py [hunk 4]
```diff
@@ -138,71 +138,58 @@ def params(self):
...
-                warnings.warn('Using qasm ast node as a circuit.Instruction '
-                              'parameter is deprecated as of the 0.11.0, and '
-                              'will be removed no earlier than 3 months after '
-                              'that release date. You should convert the qasm '
-                              'node to a supported type int, float, complex, '
-                              'str, circuit.ParameterExpression, or ndarray '
-                              'before setting Instruction.parameters',
+            self._params.append(self._normalize_parameter(single_param))
+
+    def _normalize_parameter(self, parameter):
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: I renamed the method to `validate_parameter`... and documentation added.

File: qiskit/circuit/gate.py (hunk 3 above)
Comment (1ucian0) [reply qiskit/circuit/gate.py]: I fully changed it... have another look!

File: qiskit/circuit/instruction.py [hunk 5]
```diff
@@ -137,29 +137,11 @@ def params(self):
...
-            elif isinstance(single_param, numpy.number):
-                self._params.append(single_param.item())
-            else:
-                raise CircuitError("invalid param type {0} in instruction "
-                                   "{1}".format(type(single_param), self.name))
+            self._params.append(self.validate_parameter(single_param))
+
+    def validate_parameter(self, parameter):
+        """Instruction parameters has no validation or normalization."""
+        return parameter
```
Comment (mtreinish) [inline qiskit/circuit/instruction.py:144]: Is there a reason to remove the strict type checking here as a fallback, especially for all other downstream subclasses here? This is a bit of a change in behavior and we do have subclass users of the instruction class outside of terra today. The easiest example of the top of my head is aer's snapshot instruction (which has several subclasses too): https://github.com/Qiskit/qiskit-aer/blob/master/qiskit/providers/aer/extensions/snapshot.py#L25

**Comment by mtreinish** [comment]
Oh, I just came up with a fun example:

```
from qiskit import *
from qiskit.extensions import *
from qiskit.providers.basicaer import BasicAer
                         
sim = BasicAer.get_backend('qasm_simulator')
qc = QuantumCircuit(2)
qc.h(0)       
qc.cx(0, 1)   
              
barrier = Barrier(2)
barrier.params = qc
qc.append(barrier, [0, 1])
print(qc)
print(execute(qc, sim).result().get_counts())
```

Running this with this PR branch results in a Qobj jsonschema validation error in basic aer because it has a quantum circuit object in the qobj for a parameter. If you run that on master it fails trying to set params with a `CircuitError`. Not that this is really a sane example, but I think that reinforces my thinking on why the checking was there, because of limitations on the output format from circuits for running. Personally I don't think I really have convinced myself one way or the other.

**Comment by 1ucian0** [comment]
57854df24 adds `validate_parameter` to barrier.

File: qiskit/circuit/instruction.py (hunk 5 above)
Comment (mtreinish) [reply qiskit/circuit/instruction.py:144]: As was discussed offline the conclusion to this was that it's the responsibility of users to set a sane parameter for opaque instructions or for subclasses of opaque instructions (if `validate_parameter` is not defined for the subclass)

**Pull request merged**